### ISSUE TITLE
strange pattern match

### ISSUE THREAD

**Issue by tonymorris** [data61/fp-course#178]
`FileIO.hs`

This code fails to compile, confusing `[]` with `List`.

```
do x:.y <- getArgs; return x
```

**Comment by MIJOTHY** [comment]
I ran across the [same issue](https://www.reddit.com/r/haskellquestions/comments/8o3mgm/confusing_typeerror_on_nonexhaustive_pattern/) (code and warnings/error for a bit more context below)
```haskell
main2 :: IO ()
main2 = do
  paths <- getArgs
  case paths of
    (path :. _) -> run path
    _           -> putStrLn "usage: :main \"share/files.txt\""

main3 :: IO ()
main3 = do
  paths <- getArgs
  case paths of
    (path :. _) -> run path

main4 :: IO ()
main4 = do
  (path :. _) <- getArgs
  run path
```

main3:
```
FileIO.hs|144 col 3 warning| Pattern match(es) are non-exhaustive
|| In a case alternative: Patterns not matched: Nil
```

main4:
```
FileIO.hs|143 col 3 error| • Couldn't match type ‘[Char]’ with ‘List Char’
||   Expected type: Chars
||     Actual type: [Char]
|| • In a stmt of a 'do' block: (path :. _) <- getArgs
||   In the expression:
||     do { (path :. _) <- getArgs;
||          run path }
||   In an equation for ‘main3’:
||       main3
||         = do { (path :. _) <- getArgs;
||                run path }
```

A user on the qfpl IRC (dibblego I believe) mentioned it may have something to do with the irrefutable pattern-match resulting in `fail` for the monad instance being called (excuse my likely butchering of their answer).    

After a bit of investigation (mainly involving the [MonadFail proposal](https://github.com/quchen/articles/blob/master/monad_fail.md), this is because of failable patterns. Let's take a look at what `main4` gets desugared to:
```haskell
main4 :: IO ()
main4 = do
  (path :. _) <- getArgs
  run path

-- becomes

main4 :: IO ()
main4 = do
  let f (path :. _) = run path
       f _ = fail "..."
  in getArgs >>= f 
```

Now I think this is the interesting bit: When we allow GHC to do the desugaring to the above, I suspect we get the wrong implementation for `fail`. Instead of `fail :: Chars -> IO b` with `type Chars = List Char` as defined in `Course.Applicative` being used, the IO monad implementation is used, which is
```haskell
fail :: [Char] -> IO b
fail = errorWithoutStackTrace
``` 
Note that if we write the desugared version ourselves, it passes compilation as it uses the 'right' `fail` definition.

There are a couple of ways to fix this:

1. Change `main4` to make it irrefutable by adding a tilde to the pattern to be matched:
```haskell
main4 :: IO ()
main4 = do
  ~(path :. _) <- getArgs :: IO (List Chars)
  run path
```
In case analysis, this desugaring doesn't occur, and instead we just get the inexhaustive warning. Once we've made the pattern irrefutable in main4, we now get the same warning as main3:
```
FileIO.hs|149 col 3 warning| Pattern match(es) are non-exhaustive
|| In a pattern binding: Patterns not matched: Nil
```

2. Wait for the aforementioned MonadFail implemention to make it into GHC, which will remove the `fail` functionality from Monad, and move it into MonadFail. 

tl;dr: GHC's desugaring of the non-exhaustive pattern-match results in a call to the 'wrong' (`IO Monad`) `fail` implementation, which causes the TypeError. If we use case statements, this desugaring doesn't occur. If we explicitly write the desugared code, the 'correct' (`Course.Applicative`) `fail` method gets called.
(I think).

**Comment by gwils** [comment]
I'm not sure what we can do about this, other than encourage students not to do this. The program should handle the empty list case properly anyway, shouldn't it?

**Comment by tonymorris** [comment]
Yeah, I think just "don't do this" is the solution.
